File: data-proxy-rust/src/congestion.rs
use crate::config::{ACTIVE_TRANSFERS, WARNING_LEVEL_FOR_BUFFER};
use serde::Serialize;
// use sysinfo::{Pid, ProcessExt, System, SystemExt};
use procfs::process::Process;
use tracing::{error, info, warn};

// use libc;

/// 代理状态结构体（对应 Go 的 ProxyStatus）
#[derive(Debug, Serialize)]
pub struct ProxyStatus {
    pub active_connections: i64,
    pub total_mem: u64,
    pub process_mem: u64,
    pub avg_cache_per_conn: f64,
    pub cache_usage_ratio: f64,
}

/// 检查拥塞状态（对应 Go 的 CheckCongestion）
pub fn check_congestion(pre: &str, all_buffer_size: usize) -> ProxyStatus {
    let mut status = ProxyStatus {
        active_connections: 0,
        total_mem: 0,
        process_mem: 0,
        avg_cache_per_conn: 0.0,
        cache_usage_ratio: 0.0,
    };

    // 当前进程已用内存 (RSS)
    status.process_mem = get_process_rss();

    // 活跃连接数
    let active = ACTIVE_TRANSFERS.load(std::sync::atomic::Ordering::SeqCst);
    if active <= 0 {
        return status;
    }
    status.active_connections = active;

    // 平均每连接内存
    let per_conn_cache = all_buffer_size * 1024; // 每连接 KB -> Bytes
    let avg_cache = status.process_mem as f64 / active as f64;
    status.avg_cache_per_conn = avg_cache;
    status.cache_usage_ratio = avg_cache / per_conn_cache as f64;

    // 日志输出 - info 风格
    info!(
        proxy_mem = status.process_mem,
        active_connections = active,
        avg_cache_per_connection = avg_cache,
        pre = pre,
        "Proxy avg cache"
    );

    // 拥塞警告
    if status.cache_usage_ratio > WARNING_LEVEL_FOR_BUFFER as f64 {
        warn!(
            WarningLevelforBuffer = WARNING_LEVEL_FOR_BUFFER,
            pre = pre,
            "Potential congestion: average per-connection buffer near 128KB"
        );
    }

    // 完整状态日志
    info!(
        pre = pre,
        status = ?status,
        "Proxy status"
    );

    status
}

// 读取【你设置的进程最大内存】（字节）
// fn get_process_max_memory() -> u64 {
//     let mut rlim = libc::rlimit {
//         rlim_cur: 0,
//         rlim_max: 0,
//     };
//
//     unsafe {
//         libc::getrlimit(libc::RLIMIT_AS, &mut rlim);
//     }
//
//     // 如果是无限制，返回 0（避免除0错误）
//     if rlim.rlim_cur == libc::RLIM_INFINITY {
//         0
//     } else {
//         rlim.rlim_cur as u64
//     }
// }

// 获取【当前进程已用内存】（字节）
// pub fn get_process_used_memory() -> u64 {
//     // 全局单例 System，避免重复创建
//     use once_cell::sync::Lazy;
//     static SYS: Lazy<tokio::sync::Mutex<sysinfo::System>> =
//         Lazy::new(|| tokio::sync::Mutex::new(sysinfo::System::new_all()));
//
//     tokio::runtime::Handle::current().block_on(async {
//         let mut sys = SYS.lock().await;
//         sys.refresh_processes();
//
//         let pid = Pid::this();
//         if let Some(process) = sys.process(pid) {
//             process.memory() * 1024 // KB → Bytes
//         } else {
//             0
//         }
//     })
// }

pub fn get_process_rss() -> u64 {
    if let Ok(proc) = Process::myself() {
        if let Ok(statm) = proc.statm() {
            // statm.resident: number of resident pages
            let page_size = procfs::page_size().unwrap_or(4096) as u64;
            return statm.resident as u64 * page_size;
        }
    }
    0
}

File: data-proxy/congestion/congestion.go
package congestion

import (
	"fmt"
	"github.com/gin-gonic/gin"
	"log/slog"
	"net/http"
	"runtime"
	"runtime/debug"
	"sync/atomic"
)

const (
	BufferSize            = 64
	WarningLevelforBuffer = 0.6
	//CriticalLevelforBuffer = 0.8
)

// 当前正在“真实转发数据”的请求数
var (
	ActiveTransfers int64
)

type ProxyStatus struct {
	ActiveConnections int64   `json:"active_connections"` // 当前活跃连接数
	TotalMem          int64   `json:"total_mem"`          // 机器总内存（字节）
	ProcessMem        int64   `json:"process_mem"`        // 当前进程使用内存（字节）
	AvgCachePerConn   float64 `json:"avg_cache_per_conn"` // 平均每连接缓存大小（字节）
	CacheUsageRatio   float64 `json:"cache_usage_ratio"`  // 缓存使用比例 [0,1]
}

func CheckCongestion(allBufferSize int, logger *slog.Logger) ProxyStatus {

	s := ProxyStatus{}
	// 获取系统总内存大小
	totalMem, err := getTotalMem(logger)
	if err != nil {
		logger.Error("Failed to get total memory:", err)
		return s
	}
	s.TotalMem = totalMem

	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	proxyMem := int64(m.Sys) // RSS 等价内存
	s.ProcessMem = proxyMem

	usageRatio := float64(proxyMem) / float64(totalMem)
	logger.Info(fmt.Sprintf(
		"Proxy memory: %v MiB, Total memory: %v MiB, Ratio: %.2f%%",
		proxyMem/1024/1024,
		totalMem/1024/1024,
		usageRatio*100,
	))

	//if usageRatio > WarningLevelforBuffer {
	perConnCache := allBufferSize * 1024 // 每连接 128 KB
	active := atomic.LoadInt64(&ActiveTransfers)
	if active <= 0 {
		return s
	}
	s.ActiveConnections = active

	avgCache := float64(proxyMem) / float64(active)
	logger.Info(fmt.Sprintf(
		"Active connections: %d, Average per-connection memory: %.2f KB",
		active,
		avgCache/1024,
	))

	s.AvgCachePerConn = avgCache
	s.CacheUsageRatio = avgCache / float64(perConnCache)

	if s.CacheUsageRatio > WarningLevelforBuffer {
		logger.Warn("Potential congestion: average per-connection buffer near 128KB")
	}

	logger.Info(fmt.Sprintf("Proxy status: %+v", s))
	return s
}

func getTotalMem(logger *slog.Logger) (int64, error) {

	currentLimit := debug.SetMemoryLimit(-1)
	return currentLimit, nil

	//file, err := os.Open("/proc/meminfo")
	//if err != nil {
	//	return 0, err
	//}
	//defer file.Close()
	//
	//scanner := bufio.NewScanner(file)
	//for scanner.Scan() {
	//	line := scanner.Text()
	//	// 找到 MemTotal 那一行
	//	if strings.HasPrefix(line, "MemTotal:") {
	//		parts := strings.Fields(line)
	//		if len(parts) < 3 {
	//			return 0, strconv.ErrSyntax
	//		}
	//		kb, err := strconv.ParseInt(parts[1], 10, 64)
	//		if err != nil {
	//			return 0, err
	//		}
	//		return kb * 1024, nil // 转成字节
	//	}
	//}
	//
	//return 0, os.ErrNotExist
}

// GetCongestionInfo 获取拥堵状态 /getCongestionInfo
func GetCongestionInfo(logger *slog.Logger) gin.HandlerFunc {
	return func(c *gin.Context) {
		info := CheckCongestion(2*BufferSize, logger)
		c.JSON(http.StatusOK, info)
	}
}
